README.md
# Airtable Mock

A mock library to help test Python code accessing Airtable using the [Python library](https://github.com/nicocanali/airtable-python).

It keeps tables in RAM and can do basic operations.

## Installation

The easiest way is using pip:

```sh
pip install airtablemock
```

## Usage

In your test, you patch the whole airtable library:

```py
import unittest

import airtablemock

import mycode


@airtablemock.patch(mycode.__name__ + '.airtable')
class TestMyCode(unittest.TestCase):

  def test_foo():
    # This is a client for the base "baseID", it will not access the real
    # Airtable service but only the mock one which keeps data in RAM.
    client = airtablemock.Airtable('baseID', 'apiKey')

    # Populate the table.
    client.create('table-foo', {'field1': 1, 'field2': 'two'})

    # Run your code that uses Airtable, it should transparently uses the table
    # above.
    mycode.run()

    # Access the table again to check if anything was modified.
    records = client.get('table-foo')
    …
```

## Release

To create a new release of airtablemock, update the version number in
`setup.py`, tag the Git repo and run:

```sh
python setup.py sdist upload
```


CHANGELOG.md
# Change Log

## v0.0.9 [2018-12-21]

* Add [PEP 484](https://www.python.org/dev/peps/pep-0484/) type hints.

## v0.0.7 [2018-10-12]

* Add a `create_empty_table` function.

## v0.0.6 [2018-06-26]

* Fix the `filter_by_formula` when a quoted value contains a quote.

## v0.0.5 [2018-06-22]

* Implement the `view` feature of the get and iterate method.
* Implement the `filter_by_formula` feature of the iterate method.
* Raises an error when trying to access a table that was not created before.
* Fixes the `offset` response in the get method.

## v0.0.4 [2018-06-19]

* Implement the `filter_by_formula` feature of the get method.

## v0.0.3 [2018-02-07]

* Added unit tests.
* Fixed the get method that had many syntax errors.
* Added a `clear()` method to clear all data in the mock bases.
* Added a `TestCase` class to ease independant unit testing.


test/airtablemock_test.py
"""Unit tests for the airtablemock module."""

import re
import typing
import unittest

import mock

from airtable import airtable
import airtablemock
from requests import exceptions


class AirtablemockTestCase(airtablemock.TestCase):
    """Unit tests for the Airtablemock class."""

    def test_get(self) -> None:
        """Test basic usage of the get method."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1})
        base.create('table', {'number': 2})

        records = base.get('table')['records']

        self.assertEqual([1, 2], [record['fields']['number'] for record in records])

    def test_get_by_record_id(self) -> None:
        """Test getting one record by its ID."""

        base = airtable.Airtable('base', '')
        record = base.create('table', {'number': 1})
        base.create('table', {'number': 2})

        fetched_record = base.get('table', record_id=record['id'])

        self.assertEqual(record, fetched_record)

    @mock.patch('logging.warning')
    def test_get_missing_table(self, mock_warning: mock.MagicMock) -> None:
        """Test getting a table that does not exist."""

        base = airtable.Airtable('base', '')

        match_exception = '404 .* Not Found .* {}'.format(
            re.escape('{}base/table'.format(airtable.API_URL % airtable.API_VERSION)))
        with self.assertRaisesRegex(exceptions.HTTPError, match_exception):
            base.get('table')

        mock_warning.assert_called_once()
        self.assertIn('create_empty_table', mock_warning.call_args[0][0])

    @mock.patch('logging.warning')
    def test_get_empty_table(self, mock_warning: mock.MagicMock) -> None:
        """Test getting an empty table."""

        base = airtable.Airtable('base', '')

        airtablemock.create_empty_table('base', 'table')
        base.get('table')

        mock_warning.assert_not_called()

    def test_creating_empty_table_twice(self) -> None:
        """Test creating an empty table twice."""

        airtablemock.create_empty_table('base', 'table')
        airtablemock.create_empty_table('base', 'table2')
        airtablemock.create_empty_table('base2', 'table')

        with self.assertRaises(ValueError):
            airtablemock.create_empty_table('base', 'table')

    def test_filter_by_formula_equal(self) -> None:
        """Test filtering by formula with a simple equal."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'yes'})
        base.create('table', {'number': 2, 'filter': 'no'})
        base.create('table', {'number': 3, 'filter': 'yes'})

        records = base.get('table', filter_by_formula='filter = "yes"')['records']

        self.assertEqual([1, 3], [record['fields']['number'] for record in records])

    def test_filter_by_formula_greater(self) -> None:
        """Test filtering by formula with a numerical greater than."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'yes'})
        base.create('table', {'number': 2, 'filter': 'no'})
        base.create('table', {'number': 3, 'filter': 'yes'})

        records = base.get('table', filter_by_formula='number >= 1.99')['records']

        self.assertEqual([2, 3], [record['fields']['number'] for record in records])

    def test_filter_by_formula_and(self) -> None:
        """Test filtering by formula using AND."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'yes', 'other': 'a'})
        base.create('table', {'number': 2, 'filter': 'no', 'other': 'a'})
        base.create('table', {'number': 3, 'filter': 'yes', 'other': 'b'})

        records = base.get(
            'table', filter_by_formula='AND(filter = "yes", other = "b")')['records']

        self.assertEqual([3], [record['fields']['number'] for record in records])

    def test_filter_by_formula_offset(self) -> None:
        """Test filtering by formula and using an offset."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'no'})
        base.create('table', {'number': 2, 'filter': 'yes'})
        base.create('table', {'number': 3, 'filter': 'yes'})

        # Note that the offset should be applied after the filter.
        records = base.get('table', filter_by_formula='filter = "yes"', offset=1)['records']

        self.assertEqual([3], [record['fields']['number'] for record in records])

    def test_filter_by_formula_quotes(self) -> None:
        """Test filtering by formula using a string with quotes."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': '\'"(=,)}{'})
        base.create('table', {'number': 2, 'filter': 'no'})
        base.create('table', {'number': 3, 'filter': 'yes'})

        records = base.get('table', filter_by_formula='filter = "\'\\"(=,)}{"')['records']

        self.assertEqual([1], [record['fields']['number'] for record in records])

    def test_get_limit(self) -> None:
        """Test the limit feature of the get method."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1})
        base.create('table', {'number': 2})
        base.create('table', {'number': 3})
        base.create('table', {'number': 4})
        base.create('table', {'number': 5})

        response = base.get('table', limit=2)

        self.assertEqual([1, 2], [record['fields']['number'] for record in response['records']])
        self.assertEqual(2, response.get('offset'))

        response = base.get('table', limit=2, offset=2)

        self.assertEqual([3, 4], [record['fields']['number'] for record in response['records']])
        self.assertEqual(4, response.get('offset'))

        response = base.get('table', limit=2, offset=4)

        self.assertEqual([5], [record['fields']['number'] for record in response['records']])
        self.assertNotIn('offset', response)

    def test_view(self) -> None:
        """Test calling records of a view."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'no'})
        base.create('table', {'number': 2, 'filter': 'yes'})
        base.create('table', {'number': 3, 'filter': 'yes'})

        base.create_view('table', 'filtered view', 'filter = "yes"')

        records = base.get('table', view='filtered view')['records']
        self.assertEqual([2, 3], [record['fields']['number'] for record in records])

    @mock.patch('logging.warning')
    def test_view_never_created(self, mock_logging: mock.MagicMock) -> None:
        """Test calling records of a view without ever creating any views."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'no'})
        base.create('table', {'number': 2, 'filter': 'yes'})
        base.create('table', {'number': 3, 'filter': 'yes'})

        records = base.get('table', view='filtered view')['records']
        self.assertEqual([1, 2, 3], [record['fields']['number'] for record in records])

        mock_logging.assert_called_once_with(
            'The view field is ignored as no views were created in airtablemock.')

    def test_view_and_filter(self) -> None:
        """Test filtering records of a view."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'no'})
        base.create('table', {'number': 2, 'filter': 'yes'})
        base.create('table', {'number': 3, 'filter': 'yes'})

        base.create_view('table', 'filtered view', 'filter = "yes"')

        records = base.get('table', view='filtered view', filter_by_formula='number > 2')['records']
        self.assertEqual([3], [record['fields']['number'] for record in records])

    def test_unknown_view(self) -> None:
        """Try accessing an unknown view, while views are enabled."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 1, 'filter': 'no'})

        base.create_view('table', 'existing view', 'filter = "yes"')

        match_exception = '422 .* Unprocessable Entity .* {}'.format(
            re.escape('{}base/table?view=non%20existing%20view'.format(
                airtable.API_URL % airtable.API_VERSION)))
        with self.assertRaisesRegex(exceptions.HTTPError, match_exception):
            base.get('table', view='non existing view')

    def test_iterate(self) -> None:
        """Test basic usage of the iterate method."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 3})
        base.create('table', {'number': 4})

        records = base.iterate('table')

        self.assertEqual([3, 4], [record['fields']['number'] for record in records])

    def test_iterate_filter_by_formula(self) -> None:
        """Test filtering by formula on the iterate method."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 3})
        base.create('table', {'number': 4})

        records = base.iterate('table', filter_by_formula='number > 3')

        self.assertEqual([4], [record['fields']['number'] for record in records])

    def test_iterate_view(self) -> None:
        """Test iterating a view."""

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 3})
        base.create('table', {'number': 4})
        base.create_view('table', 'my-view', 'number < 4')

        records = base.iterate('table', view='my-view')

        self.assertEqual([3], [record['fields']['number'] for record in records])

    @mock.patch('random.randrange')
    def test_create_no_random(self, mock_randrange: mock.MagicMock) -> None:
        """Tries creating an entry without randomness."""

        mock_randrange.return_value = 14

        base = airtable.Airtable('base', '')
        base.create('table', {'number': 5})
        with self.assertRaises(RuntimeError):
            base.create('table', {'number': 6})

    def test_update(self) -> None:
        """Updates a record partially."""

        base = airtable.Airtable('base', '')
        record = base.create('table', {'number': 3, 'untouched_field': 'original'})
        record_id = record['id']
        base.update('table', record_id, {'number': 4, 'new_field': 'future'})

        fields = [record['fields'] for record in base.iterate('table')]
        self.assertEqual(
            [{'number': 4, 'untouched_field': 'original', 'new_field': 'future'}], fields)

    def test_update_all(self) -> None:
        """Updates a record entirely."""

        base = airtable.Airtable('base', '')
        record = base.create('table', {'number': 3, 'untouched_field': 'original'})
        record_id = record['id']
        base.update_all('table', record_id, {'number': 4, 'new_field': 'future'})

        fields = [record['fields'] for record in base.iterate('table')]
        self.assertEqual([{'number': 4, 'new_field': 'future'}], fields)

    def test_delete(self) -> None:
        """Delete a record."""

        base = airtable.Airtable('base', '')
        record = base.create('table', {'number': 3})
        record_id = record['id']
        base.create('table', {'number': 4})

        base.delete('table', record_id)

        records = base.iterate('table')

        self.assertEqual([4], [record['fields']['number'] for record in records])

    def test_multiple_bases(self) -> None:
        """Test using the same table name in 2 different bases."""

        base1 = airtable.Airtable('first-base', '')
        base1.create('table', {'number': 6})

        base2 = airtable.Airtable('second-base', '')
        base2.create('table', {'number': 7})

        records = base1.iterate('table')
        self.assertEqual([6], [record['fields']['number'] for record in records])

    def test_multiple_clients(self) -> None:
        """Test accessing a table from different clients."""

        base1 = airtable.Airtable('first-base', '')
        base1.create('table', {'number': 8})

        other_client_base1 = airtable.Airtable('first-base', '')
        other_client_base1.create('table', {'number': 9})

        records = base1.iterate('table')
        self.assertEqual([8, 9], [record['fields']['number'] for record in records])


class FunctionsTestCase(unittest.TestCase):
    """Tests for top level module functions."""

    def test_clear(self) -> None:
        """Basic usage of the clear function."""

        base1 = airtablemock.Airtable('first-base', '')
        base1.create('table', {'number': 8})

        airtablemock.clear()

        other_client_base1 = airtablemock.Airtable('first-base', '')
        other_client_base1.create('table', {'number': 9})

        records = base1.iterate('table')
        self.assertEqual(
            [9],
            [typing.cast(typing.Dict[str, int], record['fields'])['number'] for record in records])


if __name__ == '__main__':
    unittest.main()  # pragma: no cover


airtablemock/__init__.py
"""A mock for the airtable module to ease unittests."""

import collections
import itertools
import json
import logging
import random
import re
import sys
import unittest
from urllib import parse

import mock
from parsimonious import exceptions
from parsimonious import grammar
import requests


_API_URL = 'https://api.airtable.com/v0/'

# A dictionary of all Airtable bases accessed by MockAirtable clients.
_BASES = collections.defaultdict(lambda: collections.defaultdict(collections.OrderedDict))

# A dictionary of all views predicates grouped by base ID and table name.
_VIEWS = collections.defaultdict(lambda: collections.defaultdict(collections.OrderedDict))


def clear():
    """Drop all tables from all bases."""
    _BASES.clear()
    _VIEWS.clear()


def patch(target):
    """A function or class decorator to patch the target airtable module with this one."""
    clear()
    return mock.patch(target, new=sys.modules[__name__])


class Airtable(object):
    """Airtable mock client."""

    def __init__(self, base_id=None, api_key=None):
        self.base_id = base_id
        self.api_key = api_key

    def _table(self, table_name, must_exist=True):
        if must_exist and table_name not in _BASES[self.base_id]:
            logging.warning(
                'Testers, before accessing a table it should be created, probably in your '
                'test code. Either:\n'
                ' - insert a record using\n'
                '   airtable.Airtable({}).create({}, record)\n'
                ' - create an empty table using\n'
                '   airtablemock.create_empty_table({}, {})'.format(
                    repr(self.base_id), repr(table_name), repr(self.base_id), repr(table_name)))
            response = requests.Response()
            response.status_code = 404
            response.reason = 'Not Found'
            response.url = '{}{}/{}'.format(
                _API_URL, parse.quote(self.base_id or ''), parse.quote(table_name or ''))
            response.raise_for_status()
        return _BASES[self.base_id][table_name]

    def iterate(self, table_name, batch_size=0, filter_by_formula=None, view=None):
        """Iterate over all records of a table."""
        if batch_size:
            logging.info('batch_size ignored in MockAirtableClient.iterate')

        items = self._table(table_name).items()

        if view:
            if _VIEWS:
                view_predicate = _VIEWS[self.base_id][table_name].get(view)
                if not view_predicate:
                    response = requests.Response()
                    response.status_code = 422
                    response.reason = 'Unprocessable Entity'
                    response.url = '{}{}/{}?view={}'.format(
                        _API_URL, parse.quote(self.base_id or ''), parse.quote(table_name or ''),
                        parse.quote(view))
                    response.raise_for_status()
                items = filter(view_predicate, items)
            else:
                logging.warning(
                    'The view field is ignored as no views were created in airtablemock.')

        if filter_by_formula:
            items = filter(_create_predicate(filter_by_formula), items)

        for key, fields in items:
            yield {'id': key, 'fields': fields}

    def get(self, table_name, record_id=None, limit=0, offset=None,
            filter_by_formula=None, view=None):
        """Get a list of records from a table.

        The view parameter is handled specifically compared to the real API: by
        default it's just ignored (the full table is used), however as soon as
        the create_view method is used, the view must exists or this function
        will return an error (the same error that Airtable would respond in
        case of a incorrect view).
        """
        table = self._table(table_name)

        if record_id:
            return {'id': record_id, 'fields': self._table(table_name)[record_id]}

        items = self.iterate(table_name, filter_by_formula=filter_by_formula, view=view)

        if offset:
            items = itertools.islice(items, offset, None)

        if not limit or limit > 100:
            # Default value, on Airtable server.
            limit = 100
        items = itertools.islice(items, limit)

        all_items = list(items)
        response = {'records': all_items}
        if len(all_items) + (offset or 0) < len(table):
            response['offset'] = (offset or 0) + len(all_items)
        return response

    def create(self, table_name, data):
        """Create a new record."""
        table = self._table(table_name, must_exist=False)
        for unused_i in range(30):
            record_id = _generate_random_id()
            if record_id not in table:
                break
        else:
            raise RuntimeError('Could not generate a new random ID')

        table[record_id] = data
        return {'id': record_id, 'fields': data}

    def update(self, table_name, record_id, data):
        """Update one record partially."""
        table = self._table(table_name)
        table[record_id].update(data)
        return {'id': record_id, 'fields': table[record_id]}

    def update_all(self, table_name, record_id, data):
        """Update one record completely."""
        table = self._table(table_name)
        table[record_id] = data
        return {'id': record_id, 'fields': table[record_id]}

    def delete(self, table_name, record_id):
        """Delete a record."""
        table = self._table(table_name)
        del table[record_id]
        return {'id': record_id, 'deleted': True}

    def create_view(self, table_name, view_name, formula):
        """Creates a view on a given table.

        This is not part of the official API, so you should only use this in tests.
        """
        # TODO(pascal): Implement the different sorting.
        _VIEWS[self.base_id][table_name][view_name] = _create_predicate(formula)


# See grammar at
# https://support.airtable.com/hc/en-us/articles/203255215-Formula-field-reference
_FORMULA_GRAMMAR = grammar.Grammar(
    r'''
    expression  = simple_expression / function_call
    simple_expression = field_or_value blank operator blank field_or_value
    function_call = binary_function "(" expression "," blank expression ")"
    field_or_value = field / numeric / text
    blank       = ~"\s*"
    field       = ~"[a-z_]\w*"i
    operator    = "=" / "!=" / "<=" / ">=" / "<" / ">"
    numeric     = "-"? positive_number ("." positive_number)?
    positive_number = ~"[0-9]+"
    text        = "\"" quoted_text "\""
    quoted_text = ~"([^\"\\\\]|\\\\.)*"
    binary_function = "AND" / "OR"
    '''
)


def _create_predicate(formula):
    try:
        formula_parsed = _FORMULA_GRAMMAR.parse(formula)
    except exceptions.ParseError:
        raise NotImplementedError(
            'The filter_by_formula feature is not implemented in airtablemock for this formula {}.'
            .format(formula))
    return _create_predicate_from_node(formula_parsed)


def _create_predicate_from_node(formula):
    if formula.expr_name == 'expression' and len(formula.children) == 1:
        formula = formula.children[0]

    if formula.expr_name == 'simple_expression' and len(formula.children) == 5:
        operator = formula.children[2].text
        get_a = _create_value_getter_from_node(formula.children[0])
        get_b = _create_value_getter_from_node(formula.children[-1])
        if operator == '=':
            return lambda *args: get_a(*args) == get_b(*args)
        if operator == '!=':
            return lambda *args: get_a(*args) != get_b(*args)
        if operator == '<':
            return lambda *args: get_a(*args) < get_b(*args)
        if operator == '<=':
            return lambda *args: get_a(*args) <= get_b(*args)
        if operator == '>':
            return lambda *args: get_a(*args) > get_b(*args)
        if operator == '>=':
            return lambda *args: get_a(*args) >= get_b(*args)
        raise NotImplementedError(
            'Operator {} not supported yet in filter_by_formula'.format(operator))

    if formula.expr_name == 'function_call':
        funcname = formula.children[0].text
        pred1 = _create_predicate_from_node(formula.children[2])
        pred2 = _create_predicate_from_node(formula.children[5])
        if funcname == 'AND':
            return lambda *args: pred1(*args) and pred2(*args)
        if funcname == 'OR':
            return lambda *args: pred1(*args) or pred2(*args)
        raise NotImplementedError(
            'Function "{}" not supported yet in filter_by_formula'.format(funcname))

    raise NotImplementedError(
        'Grammar not supported yet in filter_by_formula: {}'.format(formula.text))


def _create_value_getter_from_node(node):
    if node.expr_name == 'field_or_value' and len(node.children) == 1:
        node = node.children[0]

    if node.expr_name == 'field':
        fieldname = node.text
        return lambda key_fields: key_fields[1].get(fieldname)

    value = json.loads(node.text)
    return lambda *args: value


class TestCase(unittest.TestCase):
    """A base class to run independant unit tests using airtablemock.

    Use this class as a base class for your test, and then access airtable as
    you would normally.
    """

    def setUp(self):
        super(TestCase, self).setUp()
        patcher = mock.patch('airtable.airtable.Airtable', Airtable)
        patcher.start()
        self.addCleanup(patcher.stop)
        self.addCleanup(clear)


def _generate_random_id():
    return 'rec%x' % random.randrange(0x10000000000)


def create_empty_table(base_id, table_name):
    """Create an empty table in the given base."""

    if table_name in _BASES[base_id]:
        raise ValueError('Table "{}" already exists in "{}"'.format(table_name, base_id))
    _BASES[base_id].__missing__(table_name)
